examples/hooks/components/QueryRuleCustomData.tsx

```diff
@@ -0,0 +1,18 @@
+import React from 'react';
+
+import { useQueryRules, UseQueryRulesProps } from 'react-instantsearch-hooks';
+import { cx } from '../cx';
+
+export type QueryRuleCustomDataProps = React.ComponentProps<'div'> &
+  UseQueryRulesProps & {
+    children: (options: { items: any[] }) => React.ReactNode;
+  };
+
+export function QueryRuleCustomData(props: QueryRuleCustomDataProps) {
```

Comment:
We can make this generic to type the items.

Reply:
I think the type is actually changing because it entirely depends on what's inside your user Data. 
I don't think we commit to any structure at the moment: `{ title: string, link: string, banner: string }` only exists in our example, but you're allowed items of different types.

NB: same choice made here: https://github.com/algolia/instantsearch.js/blob/master/src/components/QueryRuleCustomData/QueryRuleCustomData.tsx#L20

Reply:
We don't expose the component types publicly in InstantSearch.js, which is why we didn't make it generic.

At least if we provide a way for users to type them, they can deal with their own types (with unions for instance). Right now they have no way to type their data.

Reply:
It's also the connector: https://github.com/algolia/instantsearch.js/blob/master/src/connectors/query-rules/connectQueryRules.ts#L36

 To be clear, let's say:
```ts 
   userData: [{ banner: string }, { link: string }]
```

basically, we could do this

```ts
const { items } = useQueryRules<{ banner: string; } | { link: string }>(...)
// typeof items ({ banner: string; } | { link: string })[]
```

but only by introducing breaking changes on the connector types API in InstantSearch.js. Is that you meant?

Reply:
Well, we could only type the component but you're right, that would be a new concept (no BC required though), let's wait a little :)